File: Kernel Simulator/Misc/Notifiers/Notifications.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Threading

Public Module Notifications

    'Variables
    Public NotifRecents As New List(Of Notification)
    Public NotifThread As New Thread(AddressOf NotifListen) With {.Name = "Notification Thread"}
    Public NotifyUpperLeftCornerChar As String = "╔"
    Public NotifyUpperRightCornerChar As String = "╗"
    Public NotifyLowerLeftCornerChar As String = "╚"
    Public NotifyLowerRightCornerChar As String = "╝"
    Public NotifyUpperFrameChar As String = "═"
    Public NotifyLowerFrameChar As String = "═"
    Public NotifyLeftFrameChar As String = "║"
    Public NotifyRightFrameChar As String = "║"

    ''' <summary>
    ''' Notification priority
    ''' </summary>
    Public Enum NotifPriority
        ''' <summary>
        ''' Low priority. One beep.
        ''' </summary>
        Low = 1
        ''' <summary>
        ''' Medium priority. Two beeps.
        ''' </summary>
        Medium = 2
        ''' <summary>
        ''' High priority. Three beeps.
        ''' </summary>
        High = 3
    End Enum

    ''' <summary>
    ''' Notification type
    ''' </summary>
    Public Enum NotifType
        ''' <summary>
        ''' Normal notification.
        ''' </summary>
        Normal = 1
        ''' <summary>
        ''' A notification with the progress bar. Use if you're going to notify the user while your mod is doing something.
        ''' </summary>
        Progress = 2
    End Enum

    ''' <summary>
    ''' Listens for notifications and notifies the user if one has been found
    ''' </summary>
    Private Sub NotifListen()
        Dim OldNotificationsList As New List(Of Notification)(NotifRecents)
        Dim NewNotificationsList As List(Of Notification)
        While Not NotifThread.ThreadState = ThreadState.AbortRequested
            Thread.Sleep(100)
            NewNotificationsList = NotifRecents.Except(OldNotificationsList).ToList
            If NewNotificationsList.Count > 0 And Not InSaver Then
                'Update the old notifications list
                Wdbg(DebugLevel.W, "Notifications received! Recents count was {0}, Old count was {1}", NotifRecents.Count, OldNotificationsList.Count)
                OldNotificationsList = New List(Of Notification)(NotifRecents)
                EventManager.RaiseNotificationsReceived(NewNotificationsList)

                'Iterate through new notifications
                For Each NewNotification As Notification In NewNotificationsList
                    EventManager.RaiseNotificationReceived(NewNotification)

                    'Populate title and description
                    Dim Title, Desc As String
                    Dim NotifyBorderColor As ColTypes = ColTypes.LowPriorityBorder
                    Wdbg(DebugLevel.I, "Title: {0}", NewNotification.Title)
                    Wdbg(DebugLevel.I, "Desc: {0}", NewNotification.Desc)
                    Title = NewNotification.Title.Truncate(36)
                    Desc = NewNotification.Desc.Truncate(36)
                    Wdbg(DebugLevel.I, "Truncated title: {0}", Title)
                    Wdbg(DebugLevel.I, "Truncated desc: {0}", Desc)
                    Wdbg(DebugLevel.I, "Truncated title length: {0}", Title.Length)
                    Wdbg(DebugLevel.I, "Truncated desc length: {0}", Desc.Length)

                    'Set the border color
                    Select Case NewNotification.Priority
                        Case NotifPriority.Medium
                            NotifyBorderColor = ColTypes.MediumPriorityBorder
                        Case NotifPriority.High
                            NotifyBorderColor = ColTypes.HighPriorityBorder
                    End Select

                    'Write notification to console
                    Wdbg(DebugLevel.I, "Where to store: ({0}, {1}), Title top: {2}, Desc top: {3}", Console.WindowWidth - 40, Console.WindowTop, Console.WindowTop + 1, Console.WindowTop + 2)
                    WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop, True, ColTypes.Neutral)
                    WriteWhere(Title + GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 1, True, ColTypes.NotificationTitle)
                    WriteWhere(Desc + GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 2, True, ColTypes.NotificationDescription)

                    'Optionally, draw a border
                    If DrawBorderNotification Then
                        WriteWhere(NotifyUpperLeftCornerChar + NotifyUpperFrameChar.Repeat(38) + NotifyUpperRightCornerChar, Console.WindowWidth - 41, Console.WindowTop, True, NotifyBorderColor)
                        WriteWhere(NotifyLeftFrameChar, Console.WindowWidth - 41, Console.WindowTop + 1, True, NotifyBorderColor)
                        WriteWhere(NotifyLeftFrameChar, Console.WindowWidth - 41, Console.WindowTop + 2, True, NotifyBorderColor)
                        WriteWhere(NotifyLeftFrameChar, Console.WindowWidth - 41, Console.WindowTop + 3, True, NotifyBorderColor)
                        WriteWhere(NotifyRightFrameChar, Console.WindowWidth - 2, Console.WindowTop + 1, True, NotifyBorderColor)
                        WriteWhere(NotifyRightFrameChar, Console.WindowWidth - 2, Console.WindowTop + 2, True, NotifyBorderColor)
                        WriteWhere(NotifyRightFrameChar, Console.WindowWidth - 2, Console.WindowTop + 3, True, NotifyBorderColor)
                        WriteWhere(NotifyLowerLeftCornerChar + NotifyLowerFrameChar.Repeat(38) + NotifyLowerRightCornerChar, Console.WindowWidth - 41, Console.WindowTop + 4, True, NotifyBorderColor)
                    End If

                    'Beep according to priority
                    Wdbg(DebugLevel.I, "Priority: {0}", NewNotification.Priority)
                    For i As Integer = 1 To NewNotification.Priority
                        Console.Beep()
                    Next

                    'Show progress
                    If NewNotification.Type = NotifType.Progress Then
                        Do Until NewNotification.Progress >= 100 Or NewNotification.ProgressFailed
                            Dim ProgressTitle As String = Title + " (" + CStr(NewNotification.Progress) + "%)"
                            Wdbg(DebugLevel.I, "Where to store progress: {0},{1}", Console.WindowWidth - 40, Console.WindowTop + 3)
                            Wdbg(DebugLevel.I, "Progress: {0}", NewNotification.Progress)
                            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop, True, ColTypes.Neutral)
                            WriteWhere(ProgressTitle + GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 1, True, ColTypes.NotificationTitle, NewNotification.Progress)
                            WriteWhere(Desc + GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 2, True, ColTypes.NotificationDescription)
                            WriteWhere("*".Repeat(NewNotification.Progress * 100 / 100 * (38 / 100)), Console.WindowWidth - 40, Console.WindowTop + 3, True, ColTypes.NotificationProgress)
                            Thread.Sleep(1)
                            If NewNotification.ProgressFailed Then WriteWhere(ProgressTitle + GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 1, True, ColTypes.NotificationFailure, NewNotification.Progress)
                        Loop
                    End If

                    'Clear the area
                    Dim TopTitleClear As Integer = Console.WindowTop + 1
                    Dim TopDescClear As Integer = Console.WindowTop + 2
                    Dim TopProgClear As Integer = Console.WindowTop + 3
                    Dim TopOpenBorderClear As Integer = Console.WindowTop
                    Dim TopCloseBorderClear As Integer = Console.WindowTop + 4
                    Thread.Sleep(5000)
                    NotifClearArea(Console.WindowWidth - If(DrawBorderNotification, 41, 40), TopTitleClear, TopDescClear, TopProgClear, TopOpenBorderClear, TopCloseBorderClear)
                Next
            End If
        End While
    End Sub

    ''' <summary>
    ''' Clears the area of the displayed notification
    ''' </summary>
    ''' <param name="Width">Console width</param>
    ''' <param name="TopTitle">Vertical location of title</param>
    ''' <param name="TopDesc">Vertical location of description</param>
    Private Sub NotifClearArea(Width As Integer, TopTitle As Integer, TopDesc As Integer, TopProg As Integer, TopOpenBorder As Integer, TopCloseBorder As Integer)
        If DrawBorderNotification Then
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 41, Console.WindowTop, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopOpenBorder, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 41, Console.WindowTop + 1, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopTitle, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 41, Console.WindowTop + 2, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopDesc, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 41, Console.WindowTop + 3, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopProg, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 41, Console.WindowTop + 4, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopCloseBorder, True, ColTypes.Neutral)
        Else
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 1, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopTitle, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 2, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopDesc, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Console.WindowWidth - 40, Console.WindowTop + 3, True, ColTypes.Neutral)
            WriteWhere(GetEsc() + "[0K", Width, TopProg, True, ColTypes.Neutral)
        End If
    End Sub

    ''' <summary>
    ''' Sends notification
    ''' </summary>
    ''' <param name="notif">Instance of notification holder</param>
    Public Sub NotifySend(notif As Notification)
        Wdbg(DebugLevel.I, "List contains this notification? {0}", NotifRecents.Contains(notif))
        If Not NotifRecents.Contains(notif) Then
            NotifRecents.Add(notif)
            EventManager.RaiseNotificationSent(notif)
        End If
    End Sub

    ''' <summary>
    ''' Sends notifications
    ''' </summary>
    ''' <param name="notifs">Instances of notification holder</param>
    Public Sub NotifySendRange(notifs As List(Of Notification))
        For Each notif As Notification In notifs
            NotifySend(notif)
        Next
        EventManager.RaiseNotificationsSent(notifs)
    End Sub

    ''' <summary>
    ''' Dismisses notification
    ''' </summary>
    ''' <param name="ind">Index of notification</param>
    Public Function NotifDismiss(ind As Integer) As Boolean
        Try
            NotifRecents.RemoveAt(ind)
            Wdbg(DebugLevel.I, "Removed index {0} from notification list", ind)
            EventManager.RaiseNotificationDismissed()
            Return True
        Catch ex As Exception
            Wdbg(DebugLevel.E, "Error trying to dismiss notification: {0}", ex.Message)
            WStkTrc(ex)
        End Try
        Return False
    End Function

End Module

File: public/Kernel Simulator/Misc/Notifications/Notifications.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Extensification.StringExts;
using KS.ConsoleBase.Colors;
using KS.Kernel;
using KS.Kernel.Debugging;
using KS.Misc.Splash;
using KS.Misc.Text;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.FancyWriters;

namespace KS.Misc.Notifications
{
    /// <summary>
    /// Notifications module
    /// </summary>
    public static class Notifications
    {

        /// <summary>
        /// Recent notifications
        /// </summary>
        public static List<Notification> NotifRecents = new();
        /// <summary>
        /// Notification listener thread
        /// </summary>
        public static KernelThread NotifThread = new("Notification Thread", false, NotifListen);
        /// <summary>
        /// Upper left corner character for the notfication box
        /// </summary>
        public static string NotifyUpperLeftCornerChar = "╔";
        /// <summary>
        /// Upper right corner character for the notfication box
        /// </summary>
        public static string NotifyUpperRightCornerChar = "╗";
        /// <summary>
        /// Lower left corner character for the notfication box
        /// </summary>
        public static string NotifyLowerLeftCornerChar = "╚";
        /// <summary>
        /// Lower right corner character for the notfication box
        /// </summary>
        public static string NotifyLowerRightCornerChar = "╝";
        /// <summary>
        /// Upper frame character for the notfication box
        /// </summary>
        public static string NotifyUpperFrameChar = "═";
        /// <summary>
        /// Lower frame character for the notfication box
        /// </summary>
        public static string NotifyLowerFrameChar = "═";
        /// <summary>
        /// Left frame character for the notfication box
        /// </summary>
        public static string NotifyLeftFrameChar = "║";
        /// <summary>
        /// Right frame character for the notfication box
        /// </summary>
        public static string NotifyRightFrameChar = "║";

        /// <summary>
        /// Notification priority
        /// </summary>
        public enum NotifPriority
        {
            /// <summary>
            /// Low priority. One beep.
            /// </summary>
            Low = 1,
            /// <summary>
            /// Medium priority. Two beeps.
            /// </summary>
            Medium = 2,
            /// <summary>
            /// High priority. Three beeps.
            /// </summary>
            High = 3,
            /// <summary>
            /// Custom priority. Custom colors, beeps, etc.
            /// </summary>
            Custom = 4
        }

        /// <summary>
        /// Notification type
        /// </summary>
        public enum NotifType
        {
            /// <summary>
            /// Normal notification.
            /// </summary>
            Normal = 1,
            /// <summary>
            /// A notification with the progress bar. Use if you're going to notify the user while your mod is doing something.
            /// </summary>
            Progress = 2
        }

        /// <summary>
        /// Listens for notifications and notifies the user if one has been found
        /// </summary>
        private static void NotifListen()
        {
            try
            {
                var OldNotificationsList = new List<Notification>(NotifRecents);
                List<Notification> NewNotificationsList;
                while (!Flags.KernelShutdown)
                {
                    Thread.Sleep(100);
                    NewNotificationsList = NotifRecents.Except(OldNotificationsList).ToList();
                    if (NewNotificationsList.Count > 0 & !Screensaver.Screensaver.InSaver)
                    {
                        // Update the old notifications list
                        DebugWriter.WriteDebug(DebugLevel.W, "Notifications received! Recents count was {0}, Old count was {1}", NotifRecents.Count, OldNotificationsList.Count);
                        OldNotificationsList = new List<Notification>(NotifRecents);
                        Kernel.Events.EventsManager.FireEvent("NotificationsReceived", NewNotificationsList);

                        // Iterate through new notifications. If we're on the booting stage, ensure that the notifications are only queued until the
                        // kernel has finished booting.
                        while (!SplashReport.KernelBooted)
                            Thread.Sleep(100);
                        foreach (Notification NewNotification in NewNotificationsList)
                        {
                            Kernel.Events.EventsManager.FireEvent("NotificationReceived", NewNotification);

                            // Populate title and description
                            string Title, Desc;
                            DebugWriter.WriteDebug(DebugLevel.I, "Title: {0}", NewNotification.Title);
                            DebugWriter.WriteDebug(DebugLevel.I, "Desc: {0}", NewNotification.Desc);
                            Title = NewNotification.Title.Truncate(36);
                            Desc = NewNotification.Desc.Truncate(36);
                            DebugWriter.WriteDebug(DebugLevel.I, "Truncated title: {0}", Title);
                            DebugWriter.WriteDebug(DebugLevel.I, "Truncated desc: {0}", Desc);
                            DebugWriter.WriteDebug(DebugLevel.I, "Truncated title length: {0}", Title.Length);
                            DebugWriter.WriteDebug(DebugLevel.I, "Truncated desc length: {0}", Desc.Length);

                            // Set the border color
                            DebugWriter.WriteDebug(DebugLevel.I, "Priority: {0}", NewNotification.Priority);
                            var NotifyBorderColor = ColorTools.GetColor(ColorTools.ColTypes.LowPriorityBorder);
                            var NotifyTitleColor = ColorTools.GetColor(ColorTools.ColTypes.NotificationTitle);
                            var NotifyDescColor = ColorTools.GetColor(ColorTools.ColTypes.NotificationDescription);
                            var NotifyProgressColor = ColorTools.GetColor(ColorTools.ColTypes.NotificationProgress);
                            var NotifyProgressFailureColor = ColorTools.GetColor(ColorTools.ColTypes.NotificationFailure);
                            switch (NewNotification.Priority)
                            {
                                case NotifPriority.Medium:
                                    {
                                        NotifyBorderColor = ColorTools.GetColor(ColorTools.ColTypes.MediumPriorityBorder);
                                        break;
                                    }
                                case NotifPriority.High:
                                    {
                                        NotifyBorderColor = ColorTools.GetColor(ColorTools.ColTypes.HighPriorityBorder);
                                        break;
                                    }
                                case NotifPriority.Custom:
                                    {
                                        NotifyBorderColor = NewNotification.CustomColor;
                                        NotifyTitleColor = NewNotification.CustomTitleColor;
                                        NotifyDescColor = NewNotification.CustomDescriptionColor;
                                        NotifyProgressColor = NewNotification.CustomProgressColor;
                                        NotifyProgressFailureColor = NewNotification.CustomProgressFailureColor;
                                        break;
                                    }
                            }
                            if (NewNotification.NotificationBorderColor is not null)
                            {
                                NotifyBorderColor = NewNotification.NotificationBorderColor;
                            }

                            // Write notification to console
                            DebugWriter.WriteDebug(DebugLevel.I, "Where to store: ({0}, {1}), Title top: {2}, Desc top: {3}", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop, ConsoleBase.ConsoleWrapper.WindowTop + 1, ConsoleBase.ConsoleWrapper.WindowTop + 2);
                            TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop, true, ColorTools.ColTypes.NeutralText);
                            TextWriterWhereColor.WriteWhere(Title + Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, NotifyTitleColor);
                            TextWriterWhereColor.WriteWhere(Desc + Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, NotifyDescColor);
                            TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 3, true, ColorTools.ColTypes.NeutralText);

                            // Optionally, draw a border
                            if (Flags.DrawBorderNotification)
                            {
                                // Prepare the variables
                                string CurrentNotifyUpperLeftCornerChar = NotifyUpperLeftCornerChar;
                                string CurrentNotifyUpperRightCornerChar = NotifyUpperRightCornerChar;
                                string CurrentNotifyLowerLeftCornerChar = NotifyLowerLeftCornerChar;
                                string CurrentNotifyLowerRightCornerChar = NotifyLowerRightCornerChar;
                                string CurrentNotifyUpperFrameChar = NotifyUpperFrameChar;
                                string CurrentNotifyLowerFrameChar = NotifyLowerFrameChar;
                                string CurrentNotifyLeftFrameChar = NotifyLeftFrameChar;
                                string CurrentNotifyRightFrameChar = NotifyRightFrameChar;

                                // Get custom corner characters
                                if (NewNotification.Priority == NotifPriority.Custom)
                                {
                                    CurrentNotifyUpperLeftCornerChar = NewNotification.CustomUpperLeftCornerChar;
                                    CurrentNotifyUpperRightCornerChar = NewNotification.CustomUpperRightCornerChar;
                                    CurrentNotifyLowerLeftCornerChar = NewNotification.CustomLowerLeftCornerChar;
                                    CurrentNotifyLowerRightCornerChar = NewNotification.CustomLowerRightCornerChar;
                                    CurrentNotifyUpperFrameChar = NewNotification.CustomUpperFrameChar;
                                    CurrentNotifyLowerFrameChar = NewNotification.CustomLowerFrameChar;
                                    CurrentNotifyLeftFrameChar = NewNotification.CustomLeftFrameChar;
                                    CurrentNotifyRightFrameChar = NewNotification.CustomRightFrameChar;
                                }

                                // Just draw the border!
                                TextWriterWhereColor.WriteWhere(CurrentNotifyUpperLeftCornerChar + CurrentNotifyUpperFrameChar.Repeat(38L) + CurrentNotifyUpperRightCornerChar, ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyLeftFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyLeftFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyLeftFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 3, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyRightFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 2, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyRightFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 2, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyRightFrameChar, ConsoleBase.ConsoleWrapper.WindowWidth - 2, ConsoleBase.ConsoleWrapper.WindowTop + 3, true, NotifyBorderColor);
                                TextWriterWhereColor.WriteWhere(CurrentNotifyLowerLeftCornerChar + CurrentNotifyLowerFrameChar.Repeat(38L) + CurrentNotifyLowerRightCornerChar, ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 4, true, NotifyBorderColor);
                            }

                            // Beep according to priority
                            int BeepTimes = (int)NewNotification.Priority;
                            if (NewNotification.Priority == NotifPriority.Custom)
                                BeepTimes = NewNotification.CustomBeepTimes;
                            for (int i = 1, loopTo = BeepTimes; i <= loopTo; i++)
                                ConsoleBase.ConsoleWrapper.Beep();

                            // Show progress
                            if (NewNotification.Type == NotifType.Progress)
                            {
                                while (!(NewNotification.Progress >= 100 | NewNotification.ProgressFailed))
                                {
                                    string ProgressTitle = Title + " (" + NewNotification.Progress.ToString() + "%)";
                                    DebugWriter.WriteDebug(DebugLevel.I, "Where to store progress: {0},{1}", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 3);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Progress: {0}", NewNotification.Progress);
                                    TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop, true, ColorTools.ColTypes.NeutralText);
                                    TextWriterWhereColor.WriteWhere(ProgressTitle + Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, NotifyTitleColor, NewNotification.Progress);
                                    TextWriterWhereColor.WriteWhere(Desc + Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, NotifyDescColor);
                                    TargetedProgressBarColor.WriteProgressTargeted(NewNotification.Progress, ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 3, 36, NotifyProgressColor, NotifyBorderColor, Flags.DrawBorderNotification);
                                    Thread.Sleep(1);
                                    if (NewNotification.ProgressFailed)
                                        TextWriterWhereColor.WriteWhere(ProgressTitle + Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, NotifyProgressFailureColor, NewNotification.Progress);
                                }
                            }

                            // Clear the area
                            int TopTitleClear = ConsoleBase.ConsoleWrapper.WindowTop + 1;
                            int TopDescClear = ConsoleBase.ConsoleWrapper.WindowTop + 2;
                            int TopProgClear = ConsoleBase.ConsoleWrapper.WindowTop + 3;
                            int TopOpenBorderClear = ConsoleBase.ConsoleWrapper.WindowTop;
                            int TopCloseBorderClear = ConsoleBase.ConsoleWrapper.WindowTop + 4;
                            Thread.Sleep(5000);
                            NotifClearArea(ConsoleBase.ConsoleWrapper.WindowWidth - (Flags.DrawBorderNotification ? 41 : 40), TopTitleClear, TopDescClear, TopProgClear, TopOpenBorderClear, TopCloseBorderClear);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Shutting down botification thread because of {0}", ex.Message);
                DebugWriter.WriteDebugStackTrace(ex);
            }
        }

        /// <summary>
        /// Clears the area of the displayed notification
        /// </summary>
        /// <param name="Width">Console width</param>
        /// <param name="TopTitle">Vertical location of title</param>
        /// <param name="TopDesc">Vertical location of description</param>
        /// <param name="TopProg">Vertical location of progress</param>
        /// <param name="TopOpenBorder">Vertical location of open border</param>
        /// <param name="TopCloseBorder">Vertical location of close border</param>
        private static void NotifClearArea(int Width, int TopTitle, int TopDesc, int TopProg, int TopOpenBorder, int TopCloseBorder)
        {
            if (Flags.DrawBorderNotification)
            {
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopOpenBorder, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopTitle, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopDesc, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 3, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopProg, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 41, ConsoleBase.ConsoleWrapper.WindowTop + 4, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopCloseBorder, true, ColorTools.ColTypes.NeutralText);
            }
            else
            {
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 1, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopTitle, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 2, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopDesc, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", ConsoleBase.ConsoleWrapper.WindowWidth - 40, ConsoleBase.ConsoleWrapper.WindowTop + 3, true, ColorTools.ColTypes.NeutralText);
                TextWriterWhereColor.WriteWhere(Convert.ToString(CharManager.GetEsc()) + "[0K", Width, TopProg, true, ColorTools.ColTypes.NeutralText);
            }
        }

        /// <summary>
        /// Sends notification
        /// </summary>
        /// <param name="notif">Instance of notification holder</param>
        public static void NotifySend(Notification notif)
        {
            DebugWriter.WriteDebug(DebugLevel.I, "List contains this notification? {0}", NotifRecents.Contains(notif));
            if (!NotifRecents.Contains(notif))
            {
                NotifRecents.Add(notif);
                Kernel.Events.EventsManager.FireEvent("NotificationSent", notif);
            }
        }

        /// <summary>
        /// Sends notifications
        /// </summary>
        /// <param name="notifs">Instances of notification holder</param>
        public static void NotifySendRange(List<Notification> notifs)
        {
            foreach (Notification notif in notifs)
                NotifySend(notif);
            Kernel.Events.EventsManager.FireEvent("NotificationsSent", notifs);
        }

        /// <summary>
        /// Dismisses notification
        /// </summary>
        /// <param name="ind">Index of notification</param>
        public static bool NotifDismiss(int ind)
        {
            try
            {
                NotifRecents.RemoveAt(ind);
                DebugWriter.WriteDebug(DebugLevel.I, "Removed index {0} from notification list", ind);
                Kernel.Events.EventsManager.FireEvent("NotificationDismissed");
                return true;
            }
            catch (Exception ex)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Error trying to dismiss notification: {0}", ex.Message);
                DebugWriter.WriteDebugStackTrace(ex);
            }
            return false;
        }

    }
}
